### ISSUE TITLE
PTO2 fanin spill pool deadlock on whole-tensor reads after disjoint view writes

### ISSUE THREAD

**Issue by zhaozhaozz** [ChaoWao/simpler#729]
## Summary

This issue has two related layers:

1. **Fatal handling / exit path**: after PTO2 detects `FATAL: Fanin Spill Pool Deadlock Detected!`, the runtime should terminate cleanly and return an actionable error to the host. It should not hang in scheduler/AICore shutdown, and should not require repeated `Ctrl-C`.
2. **Root-cause prevention**: the runtime/codegen should avoid creating the abnormal high-fanin dependency pattern that fills the fanin spill pool in the first place.

The original reproduction was seen in DeepSeek-V4 SWA decode on `a2a3sim`:

```bash
cd /home/bot/code/PTO/pypto-lib && source .venv/bin/activate
PTOAS_ROOT=/home/bot/code/PTO/ptoas-bin PYTHONPATH=.:models/deepseek/v4 \
  python models/deepseek/v4/deepseek_v4_decode_swa.py -p a2a3sim
```

The same pattern is also reproducible from `deepseek_v4_decode_swa_moe.py -p a2a3sim`.

Typical fatal log from the full SWA graph:

```text
FATAL: Fanin Spill Pool Deadlock Detected!
Fanin spill pool cannot reclaim space after 100000 spins (no progress).
  - Pool used:     16384 / 16384 (100.0%)
  - Pool top:      16385 (linear)
  - Pool tail:     1 (linear)
  - Needed:        1 entries
  - last_task_alive: 165 (stuck here)
  - current_task:    1566
  - In-flight tasks: 1401
  - Runtime env:  PTO2_RING_DEP_POOL=32768
```

## Minimal reproducer

The following reduced PyPTO script reproduces the same fanin spill-pool failure without running the full DeepSeek graph. It intentionally lowers `PTO2_RING_DEP_POOL` to make the failure small and quick.

```python
import argparse
import os

import pypto.language as pl


NUM_WRITERS = 72
NUM_CONSUMERS = 10
WIDTH = 16
DEFAULT_DEP_POOL = 64


@pl.jit
def fanin_spill_repro(
    x: pl.Tensor[[1, WIDTH], pl.FP32],
    y: pl.Out[pl.Tensor[[NUM_CONSUMERS * NUM_WRITERS, WIDTH], pl.FP32]],
):
    shared = pl.create_tensor([NUM_WRITERS, WIDTH], dtype=pl.FP32)

    for writer in pl.parallel(0, NUM_WRITERS, 1):
        with pl.at(
            level=pl.Level.CORE_GROUP,
            name_hint="fanin_repro_writer",
            optimization=pl.chunked_loop_optimizer,
        ):
            row = pl.add(x, 1.0)
            shared = pl.assemble(shared, row, [writer, 0])

    for consumer in pl.parallel(0, NUM_CONSUMERS, 1):
        y_block = pl.slice(y, [NUM_WRITERS, WIDTH], [consumer * NUM_WRITERS, 0])
        with pl.incore(name_hint="fanin_repro_whole_tensor_reader"):
            shared_tile = pl.load(shared, [0, 0], [NUM_WRITERS, WIDTH])
            out_tile = pl.add(shared_tile, 1.0)
            pl.store(out_tile, [0, 0], y_block)

    return y


def build_tensor_specs():
    import torch

    from golden import TensorSpec

    return [
        TensorSpec("x", [1, WIDTH], torch.float32, init_value=1.0),
        TensorSpec("y", [NUM_CONSUMERS * NUM_WRITERS, WIDTH], torch.float32, is_output=True),
    ]


if __name__ == "__main__":
    from golden import RunConfig, run_jit

    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-p",
        "--platform",
        type=str,
        default="a2a3sim",
        choices=["a2a3", "a2a3sim", "a5", "a5sim"],
    )
    parser.add_argument("-d", "--device", type=int, default=0)
    parser.add_argument("--dep-pool", type=int, default=DEFAULT_DEP_POOL)
    parser.add_argument("--runtime-profiling", action="store_true", default=False)
    args = parser.parse_args()

    os.environ["PTO2_RING_DEP_POOL"] = str(args.dep_pool)

    result = run_jit(
        fn=fanin_spill_repro,
        specs=build_tensor_specs(),
        golden_fn=None,
        config=RunConfig(
            compile=dict(dump_passes=True),
            runtime=dict(
                platform=args.platform,
                device_id=args.device,
                runtime_profiling=args.runtime_profiling,
            ),
        ),
    )
    if not result.passed:
        if result.error:
            print(result.error)
        raise SystemExit(1)
```

Run command:

```bash
cd /home/bot/code/PTO/pypto-lib && source .venv/bin/activate
PTOAS_ROOT=/home/bot/code/PTO/ptoas-bin PYTHONPATH=.:models/deepseek/v4 \
  python models/deepseek/v4/repro_fanin_spill_pool_deadlock.py -p a2a3sim
```

Observed reduced repro log:

```text
FATAL: Fanin Spill Pool Deadlock Detected!
Fanin spill pool cannot reclaim space after 100000 spins (no progress).
  - Pool used:     64 / 64 (100.0%)
  - Pool top:      65 (linear)
  - Pool tail:     1 (linear)
  - High water:    64
  - Needed:        1 entries
  - last_task_alive: 0 (stuck here)
  - current_task:    80
  - In-flight tasks: 80
  - Runtime env:  PTO2_RING_DEP_POOL=128
```

Generated orchestration has the dependency shape that matters:

```cpp
// 72 producers write disjoint views.
Tensor ret0__out = shared.view(... writer ...);
params_t0.add_input(ext_x);
params_t0.add_output(ret0__out);
rt_submit_aiv_task(0, params_t0);

// Each consumer reads the whole shared tensor as one input.
Tensor y_block = ext_y.view(... consumer ...);
params_t1.add_input(shared);
params_t1.add_inout(y_block);
rt_submit_aiv_task(1, params_t1);
```

## Root-cause analysis

The reduced graph mirrors the DeepSeek SWA failure pattern:

- Many producer tasks write disjoint regions/views of the same tensor.
- Later consumer tasks pass the whole tensor as one runtime input.
- Runtime TensorMap dependency lookup works on the Tensor/View memory region passed to the task. It does not know that the kernel may internally select only a small region using scalar offsets.
- Therefore, a whole-tensor input overlaps every previous disjoint view writer, so all those writers are added to the consumer fanin.
- `PTO2_FANIN_INLINE_CAP` is 64. Fanin beyond 64 goes to the fanin spill pool.
- In the full SWA graph, q-proj creates 256 disjoint writers for `q_proj_fp32`. The per-head q RMS/RoPE stage then passes the whole `q_proj_fp32` tensor and a scalar `h0`, so each head task can depend on all 256 q-proj writers.
- `PTO2FaninPool::reclaim()` reclaims spill entries only for tasks before `last_task_alive`.
- `last_task_alive` only advances through consecutive `CONSUMED` tasks. A completed old producer can still block the ring if its fanout refs have not all been released.

A temporary diagnostic run on the full SWA graph showed the stuck old producer was already completed, but not consumed:

```text
stuck task state: COMPLETED, fanin=5/5, fanout=59/85, subtasks=1/1
stuck task desc: kernels=[-1,22,-1]
```

So this is not a hardware kernel deadlock. It is a runtime resource/lifetime/backpressure failure: high-fanin task submission fills the fanin spill pool faster than the in-order `last_task_alive` based reclamation can free it.

## Layer 1: required fatal-handling fix

When either fanin or dependency pool detects a no-progress fatal condition, runtime should exit cleanly:

- `PTO2FaninPool::ensure_space()` and `PTO2DepListPool::ensure_space()` should not call `exit(1)` or spin forever.
- They should return a failure status and latch a shared fatal error, e.g. `PTO2_ERROR_DEP_POOL_OVERFLOW`.
- Orchestrator submit/allocation paths should stop submitting new work once the fatal is latched.
- Scheduler threads should observe the fatal state, broadcast exit/emergency shutdown, and write a scheduler/runtime status that host can report.
- AICore/AICPU deinit should use bounded waits for exit ACKs so shutdown cannot hang while handling the fatal path.
- Generated orchestration should not turn a failed `alloc_tensors()` / `submit_task()` into a secondary assertion such as `TaskOutputTensors::get_ref(index)`; the final host-side error should be the original runtime fatal.

Expected behavior for this layer: the program returns a deterministic runtime error promptly after the fatal log, without requiring `Ctrl-C` and without an abort caused by a secondary assertion.

## Layer 2: root-cause prevention / graph-level fix

Capacity increase is only a mitigation. The root issue is that runtime sees a whole-tensor dependency where the logical operation may read a narrow region.

Potential fixes:

1. **Prefer precise views in codegen**
   - If a generated kernel only needs a per-head or per-block region, pass that region as a Tensor view to runtime instead of passing the whole tensor plus scalar offsets.
   - For DeepSeek SWA, per-head q RMS/RoPE tasks should ideally input the relevant `q_proj_fp32` slice/view, not the full `q_proj_fp32` tensor.

2. **Add explicit region metadata when whole tensor + scalar offset is unavoidable**
   - If codegen must keep a whole tensor pointer for kernel reasons, allow the task argument to declare the actual read/write region used for dependency lookup.
   - TensorMap should then use this declared region, not the full storage span.

3. **Runtime backpressure/drain under spill-pool pressure**
   - Before declaring a fanin spill-pool deadlock, force scheduler drain/release paths and reclaim from the latest observed `last_task_alive`.
   - Add earlier pressure thresholds so orchestrator slows down high-fanin task submission before the pool is completely full.
   - This is not enough by itself for pathological graphs, but it should improve robustness.

4. **Better diagnostics and sizing hints**
   - Keep reporting pool usage, `last_task_alive`, `current_task`, and recommended `PTO2_RING_DEP_POOL`.
   - Consider logging the stuck task's state/fanout counters and the current task's fanin count when debug diagnostics are enabled.

## Acceptance criteria

- The minimal reproducer above exits cleanly with a meaningful runtime error when `PTO2_RING_DEP_POOL=64`.
- The fatal path does not hang and does not require `Ctrl-C`.
- The fatal path preserves the original error cause instead of reporting only a later assertion/abort.
- A codegen/runtime dependency-region fix prevents the DeepSeek SWA graph from filling the fanin spill pool at the current scale, or at least reduces the fanin count to the actual read region rather than all previous disjoint writers.

**Comment by ChaoWao** [comment]
- Layer 1 is a problme of this repo and fixed by #730 .
- Layer 2 shall be a user problem and fixed in other repo.
